### DISCUSSION TITLE
Can't ignore significant @Id field

### DISCUSSION THREAD

**Discussion by bjansen** [jqno/equalsverifier#785] [Q&A] [answered]
The latest version of equalsverifier complains that our class that extends `org.hibernate.envers.DefaultRevisionEntity` uses an `@Id` property as a significant field:

>java.lang.AssertionError: EqualsVerifier found a problem in class our.custom.AuditRevision.
-> Significant fields: id is marked @Id or @EmbeddedId so equals should not use it, but it does.
Suppress Warning.SURROGATE_KEY if you want to use only the @Id or @EmbeddedId field(s).

It seems to me that adding `id` to the list of ignored fields should make this error disappear, **but it does not**:

```
		EqualsVerifier.forClass(AuditRevision.class)
				.suppress(Warning.STRICT_INHERITANCE)
				.withIgnoredFields("id")
				.verify();
```

In `nl.jqno.equalsverifier.internal.checkers.fieldchecks.SignificantFieldCheck#assertFieldShouldNotBeUsed`, I see this assertion:

```java
        assertTrue(
            Formatter.of(message, fieldName),
            !fieldShouldBeIgnored ||
            !equalsChanged ||
            skipCertainTestsThatDontMatterWhenValuesAreNull
        );
```

I'm wondering if the code should read `fieldShouldBeIgnored` instead of `!fieldShouldBeIgnored`? I.e. the assertion should pass if the field should be ignored.

**Comment by jqno** [comment, marked as answer]
EqualsVerifier has recently become a bit more rigorous when it comes to JPA entities. There's a debate on whether it's better to give entities a business key (aka natural key), or a surrogate key. A business key is when all the semantically meaningful fields are part of the object's identity and assumes that the key is just there for technical reasons and doesn't actually mean anything for the user. A surrogate key is when the object's identity relies exclusively on the key, and the actual content of the fields doesn't matter.

EqualsVerifier assumes a business key by default, which means that you can't use `id` in `equals`. If you want EqualsVerifier to assume a surrogate key, you can suppress `Warning.SURROGATE_KEY` (as the error message states!), but then you can _only_ use the `id` field.

If your class has a business key but the `id` field actually has a semantic meaning, for instance if you have a Person class and the `id` field is the person's social security number, you can suppress `Warning.SURROGATE_OR_BUSINESS_KEY`, which allows you to use both `id` and other fields in `equals`.

You can read more about it in the manual: https://jqno.nl/equalsverifier/manual/jpa-entities/

`.withIgnoredFields` is intended for the opposite use case, where you have a field in your class that you're not using in `equals`. EqualsVerifier will notice that the field is not used and complain; you can tell EqualsVerifier that that's intended by having it ignored.

**Reply by bjansen** [reply]
Thanks for the explanation! I read the docs but somehow missed the `SURROGATE_OR_BUSINESS_KEY` part, which fixed my problem as expected.
